helix-core/src/indent.rs

```diff
@@ -186,103 +190,393 @@ pub fn indent_level_for_line(line: RopeSlice, tab_width: usize) -> usize {
     len / tab_width
 }
 
-/// Find the highest syntax node at position.
-/// This is to identify the column where this node (e.g., an HTML closing tag) ends.
-fn get_highest_syntax_node_at_bytepos(syntax: &Syntax, pos: usize) -> Option<Node> {
-    let tree = syntax.tree();
+/// Computes for node and all ancestors whether they are the first node on their line.
+/// The first entry in the return value represents the root node, the last one the node itself
+fn get_first_in_line(mut node: Node, byte_pos: usize, new_line: bool) -> Vec<bool> {
+    let mut first_in_line = Vec::new();
+    loop {
+        if let Some(prev) = node.prev_sibling() {
+            // If we insert a new line, the first node at/after the cursor is considered to be the first in its line
+            let first = prev.end_position().row != node.start_position().row
+                || (new_line && node.start_byte() >= byte_pos && prev.start_byte() < byte_pos);
+            first_in_line.push(Some(first));
+        } else {
+            // Nodes that have no previous siblings are first in their line if and only if their parent is
+            // (which we don't know yet)
+            first_in_line.push(None);
+        }
+        if let Some(parent) = node.parent() {
+            node = parent;
+        } else {
+            break;
+        }
+    }
 
-    // named_descendant
-    let mut node = tree.root_node().descendant_for_byte_range(pos, pos)?;
+    // The root node is by definition the first node in its line
+    first_in_line
+        .into_iter()
+        .rev()
+        .scan(true, |parent_is_first, first| {
+            *parent_is_first = first.unwrap_or(*parent_is_first);
+            Some(*parent_is_first)
+        })
+        .collect()
+}
 
-    while let Some(parent) = node.parent() {
-        if parent.start_byte() == node.start_byte() {
-            node = parent
+/// The total indent for some line of code.
+/// This is usually constructed in one of 2 ways:
+/// - Successively add indent captures to get the (added) indent from a single line
+/// - Successively add the indent results for each line
+struct IndentResult {
+    /// The total indent (the number of indent levels) is defined as max(0, indent-outdent).
+    /// The string that this results in depends on the indent style (spaces or tabs, etc.)
+    indent: usize,
+    outdent: usize,
+}
+impl IndentResult {
+    fn new() -> Self {
+        IndentResult {
+            indent: 0,
+            outdent: 0,
+        }
+    }
+    /// Add some other [IndentResult] to this.
+    /// The added indent should be the total added indent from one line
+    fn add_line(&mut self, added: &IndentResult) {
+        if added.indent > 0 && added.outdent == 0 {
+            self.indent += 1;
+        } else if added.outdent > 0 && added.indent == 0 {
+            self.outdent += 1;
+        }
```

Comment:
This seems odd to me, can't `indent` and `outdent` both be `> 0`?

Reply:
You're also not using the number from `added` and always use `1`.

Reply:
This is because of how the indentation that's added from a single line works: We can always have at most one indent or at most one outdent (because the indentation never increases by multiple levels on a single line). Here, `added.indent` and `added.outdent` will always be 1 or 0 because `added` is constructed using the function `add_capture`. I just check for `>0` instead for `==1` to prevent weird errors in case this invariant no longer holds in the future.

If there are both `indent` and `outdent` nodes on the same line, we ignore all of them, that's why there is no third case for when `added.indent > 0 && added.outdent > 0`. This is required to fix #1523.